### ISSUE TITLE
Multi-level inheritance support

### ISSUE THREAD

**Issue by suds123** [Rebilly/ReDoc#210]
Hi team

Does ReDoc supports multi-level polymorphism? I'm able to get ReDoc working for 1 level inheritance, but the second level values just appear as enum values and not as drop down list of values.
Any idea if there is no support right now and if in future it can be available?

Regards

**Comment by RomanGotsiy** [comment]
@suds123 could you share your spec or at least the minimal reproducible example.
Or at least screenshot

Thanks

**Comment by suds123** [comment]
Taking petstore example -

`definitions:
  Pet:
    type: object
    discriminator: petType
    properties:
      id:
        type: string
      name:
        type: string
      petType:
        type: string
    required:
    - name
    - petType
  Cat:
    description: A representation of a cat
    allOf:
    - $ref: '#/definitions/Pet'
    - type: object
      properties:
        coat:
          type: string
      required:
      - coat
  Dog:
    description: A representation of a dog
    discriminator: dogGroup
    allOf:
    - $ref: '#/definitions/Pet'
    - type: object
      properties:
        packSize:
          type: integer
        dogGroup:
          type: string
          enum:
          - SportingDog
          - HoundDog
      required:
      - packSize
      - dogGroup
  SportingDog:
    allOf:
    - $ref: '#/definitions/Dog'
    - type: object
      properties:
        huntingSkill:
          type: string
  HoundDog::
    allOf:
    - $ref: '#/definitions/Dog'
    - type: object
      properties:
        scentSkill:
          type: string`

**ReDoc screenshot -** 
![petstore](https://cloud.githubusercontent.com/assets/8452669/23178444/55c82b5a-f862-11e6-8414-32ad85b6c641.png)

**Comment by RomanGotsiy** [comment]
@suds123 thank you for clarification. Unfortunately this is currently not supported by ReDoc as this requires some rewrite
Will be implemented in the feature

**Comment by avkonst** [comment]
I think that this is a part of more generic issue related to the support of oneOf, anyOf, allOf. Discriminator from v2 of the spec does not make things simple and will be conflicting with the xxxOf features, once implemented. And it would be easier to get rid of the discriminator. I have removed it from my APIs and use standard JSON-schema xxxOf features. Redoc may render a nested section per each xxxOf declaration with a headline indicating type of xxx constraint and with "carousel-like" (horizontal) scroller to slide through subtype definitions from inside of the xxxOf. This would be simpler, because it would not require any schema introspection to determine discriminator-like field, would not require "select" based field and reactive response to it. The ReDoc view of sections would map one to one definitions in the JSON-schema, what would be easier to understand for API consumers. What do you think?

**Comment by Lytkini** [comment]
Hey, there!
Trying to make multi-level inheritance with this:
```yaml
product:
  type: object
  discriminator: uuid
  required:
    - group
    - name
    - uuid
  properties:
    uuid:
      $ref: '#/productUuid'
    name:
      $ref: '#/name'
    parentUuid:
      $ref: '#/parentUuid'
    group:
      $ref: '#/group'

true:
  description: Позиция является группой товаров.
  allOf:
    - $ref: '#/product'


false:
  description: Позиция является отдельным товаром.
  allOf:
    - $ref: '#/product'
    - discriminator: type
      required:
        - type
        - tax
        - measureName
        - price
      properties:
        code:
          $ref: '#/code'
        articleNumber:
          $ref: '#/articleNumber'
        barCodes:
          $ref: '#/barCodes'
        allowToSell:
          $ref: '#/allowToSell'
        description:
          $ref: '#/productDescription'
        price:
          $ref: '#/price'
        measureName:
          $ref: '#/measureName'
        tax:
          $ref: '#/tax'
        type:
          $ref: '#/productType'
```

Are there any updates about this enhancement?
Should I keep trying?

Thanks in advance.

**Comment by RomanGotsiy** [comment]
Hey @Lytkini,
no, you shouldn't keep trying. The behavior you want to achieve is not yet supported by ReDoc.
Maybe it will come together with https://github.com/Rebilly/ReDoc/issues/327

**Comment by Lytkini** [comment]
@RomanGotsiy, got it.
Thanks a lot for the quick response!

**Comment by am0nshi** [comment]
@RomanGotsiy does current library supports this functionality in 2.0.x ? We have a trouble with inheritance and params overriding due to out response object construction. In a response section override params is just ignored and shows only base one. But in a "Response samples" section response structure build ok with all proper data.

**Comment by RomanGotsiy** [comment]
@am0nshi Unfortunately, it is not supported yet.

But I think your issue is not related to this. Could you open a separate issue with a minimal reproducible sample of your issue!
Thanks

**Comment by am0nshi** [comment]
@RomanGotsiy done, could you please take a look once you have time?
https://github.com/Rebilly/ReDoc/issues/520
Demo: https://rebilly.github.io/ReDoc/?url=https://gist.githubusercontent.com/am0nshi/311c9c85489002859ae9314462870791/raw/5733da90e5462d8ba313588efea02218d0ae9182/swagger_redoc.json#

**Comment by raderio** [comment]
An ETA when it will be available?

**Comment by RomanGotsiy** [comment]
@raderio not really. This requires some clarification from OpenAPI spec authors about discriminator.

Thanks for reminding me about this issue!

I will open an issue in https://github.com/OAI/OpenAPI-Specification on weekends.

**Comment by raderio** [comment]
@RomanGotsiy can you please add the issue link here?

**Comment by raderio** [comment]
@RomanGotsiy have you open an issue in OpenAPI repository?

**Comment by RomanGotsiy** [comment]
Unfortunately not yet :(

Super busy. Will try to find time soon

**Comment by bzon** [comment]
Hopefully we can get this soon! I was following this reference https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/  for inherited request parameters and it doesn't render properly. 🎉

**Comment by doctornick42** [comment]
@RomanGotsiy 
could you please tell, is there any progress with this issue? We have to make a documentation for an existing API and we really need a multi-level discriminator support or any workaround. 

Thanks in advance.

**Comment by raderio** [comment]
> This requires some clarification from OpenAPI spec authors about discriminator.

@RomanGotsiy can you please describe what clarification do you need from OpenAPI?

**Comment by liquid207** [comment]
Any update?

**Comment by dave-yotta** [comment]
We're getting this too.  Our first level are abstract types so it's very unhelpful!

**Comment by dorrogeray** [comment]
Yep, this would be great to support! :pray:

**Comment by ilaoud** [comment]
Any update?

**Comment by dave-yotta** [comment]
Might have to be one of us that opens a PR 😆

**Comment by afmhenry** [comment]
Is there any planned support for nested oneOfs? Love redoc but this is a big feature gap

**Comment by 0x1a1a1a** [comment]
Still no updates? The issue is 7 years already

**Comment by RomanHotsiy** [comment]
Yes, there are no updates still. We remember this, but it's a low priority for us as this multi-level inheritance is quite rare. Also, some workarounds can be used to document the payload so users will understand it.

Implementing it is also quite difficult due to some design decisions in Redoc related to allOf/oneOf handling.
I am sorry it was not solved for so long but we don't have any ETA for this and I can't promise it will be solved at all.

I would be happy to review a PR that solves this, though 😉 

Cheers 🙌

**Comment by 0x1a1a1a** [comment]
Got it, thank you for the update.

> Also, some workarounds can be used to document the payload so users will understand it.

What kind of workarounds?

**Comment by RomanHotsiy** [comment]
Depends on your specific use case. Share it here and I'll suggest something.

**Comment by 0x1a1a1a** [comment]
```
  requestBody:
    content:
      application/json:
        schema:
          allOf:
            - oneOf:
              - $ref: "../properties/applicant_id.yaml"
              - $ref: "../properties/affiliated_person_id.yaml"
            - oneOf:
              - $ref: "../properties/document_1.yaml"
              - $ref: "../properties/document_2.yaml"
              - $ref: "../properties/document_3.yaml"
              - $ref: "../properties/document_4.yaml"
              - $ref: "../properties/document_5.yaml"
```

Currently, it is rendered as:
![Image](https://github.com/user-attachments/assets/4ef68a55-5af2-4583-acea-1b441b8f4062)

Thank you in advance!

**Comment by schimini** [comment]
@afmhenry @dave-yotta @liquid207  and others, do you mind sharing your examples ? Yesterday I needed this for a personal project, and I just gave it a go. (Just had to check if openapi was planning in removing discriminator because they have been thinking about it, but they moved those efforts to 4.x)

Taking this demo example for 3.1.0

```yaml
openapi: 3.1.0
   //.....

    Cat:
      description: A representation of a cat
      allOf:
        - $ref: '#/components/schemas/Pet'
        - type: object
          properties:
            huntingSkill:
              type: [string, boolean]
              description: The measured skill for hunting
              default: lazy
              example: adventurous
              enum:
                - clueless
                - lazy
                - adventurous
                - aggressive
          required:
            - huntingSkill
    Dog:
      description: A representation of a dog
      allOf:
        - $ref: '#/components/schemas/Pet'
        - type: object
          properties:
            packSize:
              type: integer
              format: int32
              description: The size of the pack the dog is from
              default: 1
              minimum: 1
            dogType:
              type: string
          discriminator: 
            propertyName: dogType
            mapping: 
              shiba: '#/components/schemas/Shiba'
              pomeranian: '#/components/schemas/Pomeranian'
          required:
            - packSize
    Shiba:
      description: The great meme
      allOf: 
        - $ref: '#/components/schemas/Dog'
        - type: object
          properties: 
            meme: 
              type: string
              enum: 
                - yes
    Pomeranian:
      description: The great meme
      allOf: 
        - $ref: '#/components/schemas/Dog'
        - type: object
          properties: 
            cute: 
              type: string
              enum: 
                - yes

    HoneyBee:
      description: A representation of a honey bee
      allOf:
        - $ref: '#/components/schemas/Pet'
        - type: object
          properties:
            honeyPerDay:
              type: number
              description: Average amount of honey produced per day in ounces
              example: 3.14
              multipleOf: .01
          required:
            - honeyPerDay
    Pet:
      type: object
      required:
        - name
        - photoUrls
        - petType
      discriminator:
        propertyName: petType
        mapping:
          cat: '#/components/schemas/Cat'
          dog: '#/components/schemas/Dog'
          bee: '#/components/schemas/HoneyBee'
      properties:
        id:
          externalDocs:
            description: 'Find more info here'
            url: 'https://example.com'
          description: Pet ID
          $ref: '#/components/schemas/Id'
        category:
          description: Categories this pet belongs to
          $ref: '#/components/schemas/Category'
        name:
          description: The name given to a pet
          type: string
          example: Guru
        photoUrls:
          description: The list of URL to a cute photos featuring pet
          type: [string, integer, 'null']
          minItems: 1
          maxItems: 10
          default: []
          xml:
            name: photoUrl
            wrapped: true
          items:
            type: string
            format: url
          if:
            x-displayName: isString
            type: string
          then:
            minItems: 1
            maxItems: 15
          else:
            x-displayName: notString
            type: [integer, 'null']
            minItems: 1
            maxItems: 20
        friend:
          $ref: '#/components/schemas/Pet'
        tags:
          description: Tags attached to the pet
          type: array
          exclusiveMaximum: 100
          exclusiveMinimum: 0
          xml:
            name: tag
            wrapped: true
          items:
            $ref: '#/components/schemas/Tag'
        status:
          type: string
          description: Pet status in the store
          enum:
            - available
            - pending
            - sold
          default: pending
        petType:
          description: Type of a pet
          type: string
        huntingSkill:
          type: [integer]
          enum:
            - 0
            - 1
            - 2
      xml:
        name: Pet
    
``` 
I was able to get this:


![Image](https://github.com/user-attachments/assets/a8da38c6-41b5-493b-a555-84ed9814c477)

![Image](https://github.com/user-attachments/assets/d981933f-1a0d-42d7-856a-ff5445651950)


@RomanHotsiy , if you're still interested in a PR I can have my spaghetti cleaned up and ready tomorrow. 
Also if you can foresee any potential edge cases let me know so I can test it properly.

Thanks

**Comment by pedrocerejo** [comment]
@RomanHotsiy any news on this?

**Comment by GuyEP-Modulate** [comment]
@RomanHotsiy Adding another voice in the chorus of those who could use this fix.

@schimini Could you prepare a PR, pretty please?

**Comment by GuyEP-Modulate** [comment]
As a workaround, I am using a script that I can run after generating the docs that will surgically remove problematic `discriminator` elements. Sadly, it also affects the copy of the schema that users can download from the documentation, but at least this enables me to keep the original schema untouched (and I can distribute that to customers separately).

The script ([remove-element-from-redocly-html.js](https://github.com/user-attachments/files/24095419/remove-element-from-redocly-html.js)) is attached. Here's how to use it:

```
node patch-redocly-html.js redoc.html \
  /components/schemas/MyType1/discriminator
  /components/schemas/MyType2/discriminator
```

Where `redoc.html` is the path to the generated Redoc output, while `/components/schemas/MyType1/discriminator` and `/components/schemas/MyType2/discriminator` are JSON pointer paths to the discriminator attributes that are breaking the Redoc display (usually, the top-most discriminator if you have two discriminators nested).

This will insert some Javascript into the HTML that removes the discriminator key(s) from the schema before Redoc gets initialized. You will need to run this each time you generate documentation and you'll need to update the path if your schema structure changes.

This is a hack/workaround, but it's better than the values just not appearing at all in the docs.
